### PR TITLE
chakrashim: add JSON methods

### PR DESCRIPTION
**Pull request by boingoing**
Add `v8::JSON::Parse` and `v8::JSON::Stringify` to chakrashim. These methods were recently added to `Nan::JSON` which a lot of native modules depend on and now fail to build against node-chakracore. One example is [this](https://github.com/sass/node-sass/pull/1777#issuecomment-309180302) thread:
https://github.com/sass/node-sass/pull/1777

Includes some new node tests in test/addons/json.

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
chakrashim

### CODE REVIEW

File: deps/chakrashim/src/v8json.cc [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+
+#include "v8.h"
+#include "jsrtutils.h"
+
+namespace v8 {
+    Local<Value> JSON::Parse(Local<String> json_string) {
```
Comment (kfarnung) [inline deps/chakrashim/src/v8json.cc]: Use two spaces for indentation of new scopes and four spaces for continuations from the previous line.

File: deps/chakrashim/src/v8json.cc [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        jsrt::IsolateShim* iso = jsrt::IsolateShim::GetCurrent();
+        jsrt::ContextShim* contextShim = iso->GetCurrentContextShim();
+
+        JsValueRef jsonParseFunction = contextShim->GetjsonParseFunction();
+        JsValueRef obj;
+        
+        if (jsrt::CallFunction(jsonParseFunction, (JsValueRef)*json_string,
+            &obj) != JsNoError) {
```
Comment (kfarnung) [inline deps/chakrashim/src/v8json.cc]: Indent this line to line up with the `jsrt::CallFunction(` parameters.

File: deps/chakrashim/src/v8json.cc [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    MaybeLocal<String> JSON::Stringify(Local<Context> context,
+        Local<Object> json_object,
+        Local<String> gap) {
+        jsrt::IsolateShim* iso = jsrt::IsolateShim::GetCurrent();
+        jsrt::ContextShim* contextShim = iso->GetCurrentContextShim();
+        
+        JsValueRef jsonStringifyFunction = contextShim->GetjsonStringifyFunction();
+        JsValueRef str;
```
Comment (kfarnung) [inline deps/chakrashim/src/v8json.cc]: Initialize to `JS_INVALID_REFERENCE`

File: deps/chakrashim/src/v8json.cc [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        JsValueRef actualGap =
+            gap.IsEmpty() ? jsrt::GetUndefined() : (JsValueRef)*gap;
```
Comment (kfarnung) [inline deps/chakrashim/src/v8json.cc]: Use C++-style cast

File: test/addons/json/binding.cc [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#include <node.h>
+#include <v8.h>
+
+namespace {
+
+void Stringify(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  auto result = v8::JSON::Stringify(args.GetIsolate()->GetCurrentContext(),
+    args[0].As<v8::Object>(),
```
Comment (kfarnung) [inline test/addons/json/binding.cc]: Indent to line up with the open paren

**Review by kfarnung — APPROVED** [review]
LGTM with minor changes

File: deps/chakrashim/src/v8json.cc [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+        return Local<Value>::New(obj);
+    }
+
+    MaybeLocal<Value> JSON::Parse(Isolate* isolate,
+        Local<String> json_string) {
+        return JSON::Parse(json_string);
+    }
+
+    MaybeLocal<Value> JSON::Parse(Local<Context> context,
```
Comment (MSLaguana) [inline deps/chakrashim/src/v8json.cc]: Given the deprecation warnings in the header, shouldn't these functions be implemented in the other order? Or is the `Local<Context>` not easily convertible to a `jsrt::ContextShim`?

File: deps/chakrashim/src/v8json.cc [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    Local<Value> JSON::Parse(Local<String> json_string) {
+        jsrt::IsolateShim* iso = jsrt::IsolateShim::GetCurrent();
+        jsrt::ContextShim* contextShim = iso->GetCurrentContextShim();
+
+        JsValueRef jsonParseFunction = contextShim->GetjsonParseFunction();
+        JsValueRef obj;
+        
+        if (jsrt::CallFunction(jsonParseFunction, (JsValueRef)*json_string,
+            &obj) != JsNoError) {
+            return Local<Value>();
```
Comment (MSLaguana) [inline deps/chakrashim/src/v8json.cc]: What is the expected behavior in the case of malformed input? It looks like the caller just gets back a null with no indication of what went wrong, is that correct?

File: deps/chakrashim/src/v8json.cc [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    MaybeLocal<Value> JSON::Parse(Local<Context> context,
+        Local<String> json_string) {
+        return JSON::Parse(json_string);
+    }
+
+    MaybeLocal<String> JSON::Stringify(Local<Context> context,
+        Local<Object> json_object,
+        Local<String> gap) {
+        jsrt::IsolateShim* iso = jsrt::IsolateShim::GetCurrent();
+        jsrt::ContextShim* contextShim = iso->GetCurrentContextShim();
```
Comment (MSLaguana) [inline deps/chakrashim/src/v8json.cc]: Again, `Local<Context>` is not something we should be using for the `jsrt::ContextShim`?

File: deps/chakrashim/src/v8json.cc (hunk 6 above)
Comment (kfarnung) [reply deps/chakrashim/src/v8json.cc]: I missed this the first time.  Ideally we should be able to take the Context and get the ContextShim, but we don't seem to have the function available.  We do have a `FromIsolate` method currently.

Comment (kfarnung) [reply deps/chakrashim/src/v8json.cc]: Found it, we should probably follow this pattern:

https://github.com/nodejs/node-chakracore/blob/76c7a22b665d0a2e6788afcab4e5bf7a5ef3ccd0/deps/chakrashim/src/v8regexp.cc#L32

File: deps/chakrashim/include/v8.h [hunk 9]
```diff
@@ -2068,6 +2069,40 @@ class V8_EXPORT SharedArrayBuffer : public Object {
...
+*/
+class V8_EXPORT JSON {
+public:
+    /**
+    * Tries to parse the string |json_string| and returns it as value if
+    * successful.
+    *
+    * \param json_string The string to parse.
+    * \return The corresponding value if successfully parsed.
+    */
```
Comment (kunalspathak) [inline deps/chakrashim/include/v8.h]: remove comments. If you see around in this file, we don't retain the comments from `deps\v8\include\v8.h` file.

File: deps/chakrashim/src/v8json.cc (hunk 7 above)
Comment (kunalspathak) [reply deps/chakrashim/src/v8json.cc]: @MSLaguana  - Yes, that's what we do today in other places, basically return an empty local.

File: deps/chakrashim/src/v8json.cc (hunk 1 above)
Comment (kunalspathak) [inline deps/chakrashim/src/v8json.cc]: You should be calling into `JSON::Parse(Local<Context> context, ....)` from here passing the current context from this API and do all the logic out there. That API should just honor whatever context was passed. See 
https://github.com/nodejs/node-chakracore/issues/95

File: deps/chakrashim/src/v8json.cc [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+#include "v8.h"
+#include "jsrtutils.h"
+
+namespace v8 {
+    Local<Value> JSON::Parse(Local<String> json_string) {
+        jsrt::IsolateShim* iso = jsrt::IsolateShim::GetCurrent();
+        jsrt::ContextShim* contextShim = iso->GetCurrentContextShim();
+
+        JsValueRef jsonParseFunction = contextShim->GetjsonParseFunction();
+        JsValueRef obj;
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8json.cc]: >obj [](start = 19, length = 3)

Initialize to `JS_INVALID_REF` ?

**Comment by kunalspathak** [comment]
Also, run cpplinter and jslinter (included as part of unit test run) which will catch most of the linting issue that @kfarnung  pointed.

File: deps/chakrashim/include/v8.h (hunk 9 above)
Comment (boingoing) [reply deps/chakrashim/include/v8.h]: Ok, thanks. Wasn't sure about this.

**Comment by boingoing** [comment]
@kunalspathak I ran the test suite but the linters failed with a message about the command line being too long. Not sure if they're just broken in general or my configuration is bad.

File: deps/chakrashim/src/v8json.cc (hunk 1 above)
Comment (boingoing) [reply deps/chakrashim/src/v8json.cc]: Is it sufficient to just get the current context via `Local<Context>()`? Seems a lot of chakrashim API do something like this:
```c++
return FromMaybe(JSON::Parse(Local<Context>(), json_string));
```

File: deps/chakrashim/src/v8json.cc (hunk 8 above)
Comment (boingoing) [reply deps/chakrashim/src/v8json.cc]: Refactored this to use the right context.

File: deps/chakrashim/src/v8json.cc (hunk 6 above)
Comment (boingoing) [reply deps/chakrashim/src/v8json.cc]: Sounds right. I reversed the order of implementation here.

File: deps/chakrashim/src/v8json.cc (hunk 1 above)
Comment (boingoing) [reply deps/chakrashim/src/v8json.cc]: I made this change and aligned the function arguments on multiple lines.

Comment (kfarnung) [reply deps/chakrashim/src/v8json.cc]: I thought that just created a default or empty `Local<T>()`?  I'm wondering if we should instead use:

https://github.com/nodejs/node-chakracore/blob/master/deps/chakrashim/src/v8context.cc#L111

File: deps/chakrashim/src/v8json.cc (hunk 2 above)
Comment (boingoing) [reply deps/chakrashim/src/v8json.cc]: Made this change here but the other call to CallFunction which has arguments spread on two lines is not lined-up because it would put that line past the limit. Doesn't seem like there's a hard policy on this?

**Comment by kfarnung** [comment]
@boingoing I was wondering about that.  It passes every file individually to the linter, maybe your added file finally pushed it over the edge.  We can probably break it up into multiple options or maybe test out wildcard usage.

File: deps/chakrashim/src/v8json.cc (hunk 3 above)
Comment (boingoing) [reply deps/chakrashim/src/v8json.cc]: Ok, good point.

File: test/addons/json/binding.cc (hunk 5 above)
Comment (boingoing) [reply test/addons/json/binding.cc]: Ok, I fixed both of these.

File: deps/chakrashim/src/v8json.cc (hunk 1 above)
Comment (boingoing) [reply deps/chakrashim/src/v8json.cc]: Changed to use:
```c++
v8::Isolate* iso = jsrt::IsolateShim::GetCurrentAsIsolate();
```
and
```c++
Local<Context> context = isolate->GetCurrentContext();
```
in the two overloads of `JSON::Parse`.

File: deps/chakrashim/src/v8json.cc (hunk 10 above)
Comment (boingoing) [reply deps/chakrashim/src/v8json.cc]: Ok, good catch, thanks.

**Review by MSLaguana — APPROVED** [review]

**Comment by boingoing** [comment]
@kfarnung and I investigated the cpplint command line a little bit and it's clear there is some combination of further work required and configuration weirdness on my machine. We agreed to clean up the lint errors in this change and come back to the cpplint command line later.

As part of this, I ran the linter on `deps/chakrashim/include/v8.h`. We normally exclude this file because it has had a lot of warnings in the past. Now, though, it doesn't have too many warnings so I have taken care of them for the most part. There are a couple more which we can't fix without breaking compatibility with v8, so I didn't change them:

```
deps\chakrashim\include\v8.h:369:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
deps\chakrashim\include\v8.h:892:  Constructors callable with one argument should be marked explicit.  [runtime/explicit] [5]
deps\chakrashim\include\v8.h:1363:  Is this a non-const reference? If so, make const or use a pointer: PropertyDescriptor& decriptor  [runtime/references] [2]
```

Adding `explicit` to the two constructors breaks compat with v8. They don't have this problem because the constructors are marked with `V8_INLINE` in the v8 version of `v8.h`. The other error can't be fixed without changing a v8 function signature - which we can't do - and, indeed, this same lint warning comes out against the real `v8.h`.

I added a commit here with the cleanups to `v8.h` but I propose we don't add it explicitly to the cpplint rule in `vcbuild.bat` until we can work out the above errors.

**Comment by kfarnung** [comment]
@boingoing you can ignore specific errors with // NOLINT(topic/rule)

See https://github.com/nodejs/node-chakracore/blob/master/deps/chakrashim/src/inspector/v8-stack-trace-impl.cc#L44

**Comment by boingoing** [comment]
@kfarnung Ok. Added `v8.h` to the cpplint rule in `vcbuild.dat` and marked those unfixable errors with `// NOLINT` in a new commit. 👍

File: deps/chakrashim/include/v8.h [hunk 11]
```diff
@@ -1765,15 +1772,15 @@ enum class ConstructorBehavior { kThrow, kAllow };
 class V8_EXPORT Function : public Object {
  public:
   static MaybeLocal<Function> New(Local<Context> context,
-                                  FunctionCallback callback,
-                                  Local<Value> data = Local<Value>(),
-                                  int length = 0,
-                                  ConstructorBehavior behavior = ConstructorBehavior::kAllow);
+    FunctionCallback callback,
```
Comment (kfarnung) [inline deps/chakrashim/include/v8.h]: super nit:  Can you bring the first parameter down and indent them all two more spaces?

**Review by kfarnung — APPROVED** [review]

**Comment by kfarnung** [comment]
@boingoing Did you kick off a CI run for this PR?

File: deps/chakrashim/include/v8.h (hunk 11 above)
Comment (boingoing) [reply deps/chakrashim/include/v8.h]: Ok, sure.

**Comment by kfarnung** [comment]
CI: https://ci.nodejs.org/job/chakracore-test-pull-request/91/

**Comment by boingoing** [comment]
The CI has a MacOS failure in the unit test I added for v8::JSON. The log on the CI isn't very useful, just says "uh oh". Tried this out on my own Mac tonight and the error I get is confusing:

```
dyld: lazy symbol binding failed: Symbol not found: __ZN2v84JSON9StringifyENS_5LocalINS_7ContextEEENS1_INS_6ObjectEEENS1_INS_6StringEEE
  Referenced from: /Users/chakradev/code/node-chakracore/test/addons/json/build/Release/binding.node
  Expected in: flat namespace

dyld: Symbol not found: __ZN2v84JSON9StringifyENS_5LocalINS_7ContextEEENS1_INS_6ObjectEEENS1_INS_6StringEEE
  Referenced from: /Users/chakradev/code/node-chakracore/test/addons/json/build/Release/binding.node
  Expected in: flat namespace

Abort trap: 6
```
Tried doing a little refactoring to make sure we're referencing the symbol the same way as other unit tests in addons but can't shake this error. The node binary I built doesn't export the symbol we're trying to dyld. I don't see any list of exports anywhere in chakrashim and it's not obvious to me why we hit this error only on MacOS.

File: deps/chakrashim/include/v8.h [hunk 12]
```diff
@@ -25,7 +25,8 @@
 // (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-#pragma once
+#ifndef DEPS_CHAKRASHIM_INCLUDE_V8_H_
+#define DEPS_CHAKRASHIM_INCLUDE_V8_H_
```
Comment (kunalspathak) [inline deps/chakrashim/include/v8.h:29]: what is the purpose of this change?

Comment (boingoing) [reply deps/chakrashim/include/v8.h:29]: #pragma causes a failure in cpplint.

Comment (kfarnung) [reply deps/chakrashim/include/v8.h:29]: Node standards use ifdef, not pragma.

**Review by kfarnung — APPROVED** [review]
LGTM

**Comment by boingoing** [comment]
New CI run:
https://ci.nodejs.org/job/chakracore-test/110

**Comment by kfarnung** [comment]
Looks like the macOS CI was successful:  https://ci.nodejs.org/job/chakracore-test-osx/114/nodes=osx1010/tapResults/

> 1660 tests, 1553 ok, **0 not ok**, 26 skipped, 81 ToDo, 0 Bail Out!

```
ok 1564 addons/json/test
  ---
  duration_ms: 0.189
  ...
```
